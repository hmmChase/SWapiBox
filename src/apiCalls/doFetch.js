export const doFetch = async url => {
  try {
    const response = await fetch(url);
    if (!response.ok) {
      throw new Error(`${response.status}`);
    }
    return await response.json();
  } catch (error) {
    throw new Error(`Network request failed. (error: ${error.message})`);
  }
};
