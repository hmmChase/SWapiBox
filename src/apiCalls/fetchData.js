import { fetchPeople } from './peopleData';
import { fetchPlanets } from './planetData';
import { fetchVehicles } from './vehicleData';

export async function fetchCategoryData(category) {
  switch (category) {
  case 'people':
    return fetchPeople(category);
  case 'planets':
    return fetchPlanets(category);
  case 'vehicles':
    return fetchVehicles(category);
  default:
    break;
  }
}

export async function makeFetch(URL) {
  try {
    const response = await fetch(URL);
    return await response.json();
  } catch (error) {
    throw new Error(error);
  }
}