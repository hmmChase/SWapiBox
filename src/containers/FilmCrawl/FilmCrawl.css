.crawl-container {
  display: flex;
  overflow: hidden;
  width: 100vw;
  height: 82vh;
  color: #feda4a;
  font-family: 'Pathway Gothic One', sans-serif;
  font-size: 8em;
  font-weight: 600;
  letter-spacing: 6px;
  perspective: 200px;
}

.crawl-content {
  position: relative;
  top: -1000px;
  transform-origin: 50% 100%;
  animation: crawl 60s linear;
  visibility: hidden;
}

.pretext {
  display: flex;
  justify-content: center;
  align-items: center;
  overflow: hidden;
  height: 80vh;
  font-size: 2em;
}

@keyframes crawl {
  0% {
    top: 1000px;
    transform: rotateX(20deg) translateZ(0);
    visibility: visible;
  }
  100% {
    top: -6000px;
    transform: rotateX(25deg) translateZ(-2000px);
  }
}
