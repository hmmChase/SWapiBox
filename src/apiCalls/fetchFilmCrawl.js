import { doFetch } from './doFetch';

export const fetchFilmCrawl = async () => {
  const root = 'https://swapi.co/api';
  const randomNum = Math.floor(Math.random() * 6 + 1);
  const randomFilmData = await doFetch(`${root}/films/${randomNum}`);
  return {
    title: randomFilmData.title,
    date: randomFilmData.release_date,
    crawl: randomFilmData.opening_crawl
  };
};