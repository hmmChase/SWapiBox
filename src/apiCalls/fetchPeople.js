import { doFetch } from './doFetch';

export const fetchPeople = async () => {
  const root = 'http://swapi.com/api';
  const peopleData = await doFetch(`${root}/people`);
  return peopleData;
};
