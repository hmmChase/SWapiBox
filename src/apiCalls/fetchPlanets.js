import { doFetch } from './doFetch';

export const fetchPlanets = async () => {
  try {
    const root = 'https://swapi.co/api';
    const planetsData = await doFetch(`${root}/planets`);
    const parsedPlanetsData = planetsData.results.map(async planet => {
      const residentsData = planet.residents.map(async residentURL => {
        return await fetchResidents(residentURL);
      });
      const resolveResidents = await Promise.all(residentsData);
      return {
        name: planet.name,
        terrain: planet.terrain,
        population: planet.population,
        climate: planet.climate,
        residents: resolveResidents.length
          ? resolveResidents.join(', ')
          : 'unknown'
      };
    });
    return await Promise.all(parsedPlanetsData);
  } catch (error) {
    throw new Error(error.message);
  }
};

export const fetchResidents = async residentURL => {
  try {
    const residentData = await doFetch(residentURL);
    return residentData.name;
  } catch (error) {
    throw new Error(error.message);
  }
};
