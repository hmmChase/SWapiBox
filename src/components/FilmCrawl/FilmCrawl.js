import React, { Component } from 'react';
import { fetchFilmCrawl } from '../../apiCalls/fetchFilmCrawl';

class FilmCrawl extends Component {
  constructor() {
    super();

    this.state = { FilmCrawl: {} };
  }

  componentDidMount = async () => {
    this.setState({ FilmCrawl: await fetchFilmCrawl() });
  };

  render() {
    return (
      <section className="crawl-container">
        <div className="crawl-content">
          <p>{this.state.FilmCrawl.crawl}</p>
          <p>{this.state.FilmCrawl.title}</p>
          <p>{this.state.FilmCrawl.date}</p>
        </div>
      </section>
    );
  }
}

export default FilmCrawl;
