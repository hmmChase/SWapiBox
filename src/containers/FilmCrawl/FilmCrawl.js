import React, { Component } from 'react';
import { fetchFilmCrawl } from '../../apiCalls/fetchFilmCrawl';
import './FilmCrawl.css';

class FilmCrawl extends Component {
  constructor(props) {
    super(props);

    this.state = {
      filmCrawl: null,
      errorStatus: null
    };
  }

  componentDidMount = () => {
    this.setFilmCrawl();
  };

  setFilmCrawl = async () => {
    try {
      this.setState({
        filmCrawl: await fetchFilmCrawl()
      });
    } catch (error) {
      this.setState({ errorStatus: error.message });
    }
  };

  renderFilmCrawl = () => {
    const { filmCrawl, errorStatus } = this.state;
    if (errorStatus) {
      return <p>{errorStatus}</p>;
    }
    if (filmCrawl) {
      return (
        <section className="crawl-container">
          <div className="crawl-content">
            <p>{this.state.filmCrawl.crawl}</p>
            <p>{this.state.filmCrawl.title}</p>
            <p>{this.state.filmCrawl.date}</p>
          </div>
        </section>
      );
    }
    return <p>...loading</p>;
  };

  render() {
    return (
      <section className="crawl-container">{this.renderFilmCrawl()}</section>
    );
  }
}

export default FilmCrawl;
