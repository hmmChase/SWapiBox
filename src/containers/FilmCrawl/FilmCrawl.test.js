import React from 'react';
import { shallow } from 'enzyme';
import FilmCrawl from './FilmCrawl';
import { fetchFilmCrawl } from '../../apiCalls/fetchFilmCrawl';
jest.mock('../../apiCalls/fetchFilmCrawl.js');

describe('FilmCrawl', () => {
  let filmCrawl;
  let mockFilmCrawlData;

  beforeEach(() => {
    jest.resetAllMocks();
    mockFilmCrawlData = {
      crawl: 'There is unrest in the Galactic Senate...',
      date: '2002-05-16',
      title: 'Attack of the Clones'
    };

    filmCrawl = shallow(<FilmCrawl />, { disableLifecycleMethods: true });
  });

  it('matches snapshot with filmCrawl data fetched', () => {
    filmCrawl.setState({
      filmCrawl: mockFilmCrawlData
    });

    expect(filmCrawl).toMatchSnapshot();
  });

  it('matches snapshot with error fetching film crawl data', () => {
    filmCrawl.setState({
      errorStatus: 'mock error'
    });

    expect(filmCrawl).toMatchSnapshot();
  });

  it('matches snapshot while fetching film crawl data', () => {
    filmCrawl = shallow(<FilmCrawl />);

    expect(filmCrawl).toMatchSnapshot();
  });

  describe('componentDidMount', () => {
    it('calls setFilmCrawl', async () => {
      // need help understanding this

      const setFilmCrawl = jest.spyOn(filmCrawl.instance(), 'setFilmCrawl');
      // setFilmCrawl = filmCrawl.instance().setFilmCrawl = jest.fn();

      filmCrawl.instance().componentDidMount();
      // filmCrawl = await shallow(<FilmCrawl />);

      expect(setFilmCrawl).toHaveBeenCalledTimes(1);
    });
  });

  describe('setFilmCrawl', () => {
    it('calls fetchFilmCrawl', async () => {
      await filmCrawl.instance().setFilmCrawl();

      expect(fetchFilmCrawl).toHaveBeenCalledTimes(1);
    });

    it('sets state.filmCrawl with fetched film crawl data,', async () => {
      fetchFilmCrawl.mockImplementation(() => mockFilmCrawlData);
      await filmCrawl.instance().setFilmCrawl();

      expect(filmCrawl.state().filmCrawl).toEqual(mockFilmCrawlData);
    });

    it('sets state.errorStatus with error if fetch fails,', async () => {
      const mockError = 'Network request failed. (error: mock)';
      fetchFilmCrawl.mockImplementation(() => {
        throw new Error(mockError);
      });
      await filmCrawl.instance().setFilmCrawl();

      expect(filmCrawl.state().errorStatus).toBe(mockError);
    });
  });

  describe('renderFilmCrawl', () => {
    it('returns error element if error set in state.errorStatus', () => {
      filmCrawl.setState({ errorStatus: 'mock error' });
      const renderFilmCrawl = filmCrawl.instance().renderFilmCrawl();

      expect(renderFilmCrawl).toEqual(<p>mock error</p>);
    });

    it('returns filmcrawl elements if filmcrawl set in state.filmCrawl', () => {
      filmCrawl.setState({ filmCrawl: mockFilmCrawlData });
      const renderFilmCrawl = filmCrawl.instance().renderFilmCrawl();

      expect(renderFilmCrawl).toEqual(
        <div>
          <p>There is unrest in the Galactic Senate...</p>
          <p>Attack of the Clones</p>
          <p>2002-05-16</p>
        </div>
      );
    });

    it('returns loading element if state.filmCrawl is null', () => {
      const renderFilmCrawl = filmCrawl.instance().renderFilmCrawl();

      expect(renderFilmCrawl).toEqual(<p>...loading</p>);
    });
  });
});
