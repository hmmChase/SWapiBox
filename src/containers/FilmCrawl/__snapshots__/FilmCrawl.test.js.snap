// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilmCrawl matches snapshot while fetching film crawl data 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FilmCrawl />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        ...loading
</p>,
      "className": "crawl-container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "...loading",
      },
      "ref": null,
      "rendered": "...loading",
      "type": "p",
    },
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          ...loading
</p>,
        "className": "crawl-container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "...loading",
        },
        "ref": null,
        "rendered": "...loading",
        "type": "p",
      },
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`FilmCrawl matches snapshot with error fetching film crawl data 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FilmCrawl />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        mock error
</p>,
      "className": "crawl-container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "mock error",
      },
      "ref": null,
      "rendered": "mock error",
      "type": "p",
    },
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          mock error
</p>,
        "className": "crawl-container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "mock error",
        },
        "ref": null,
        "rendered": "mock error",
        "type": "p",
      },
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;

exports[`FilmCrawl matches snapshot with filmCrawl data fetched 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FilmCrawl />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <section
        className="crawl-container"
>
        <div
                className="crawl-content"
        >
                <p>
                        There is unrest in the Galactic Senate...
                </p>
                <p>
                        Attack of the Clones
                </p>
                <p>
                        2002-05-16
                </p>
        </div>
</section>,
      "className": "crawl-container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="crawl-content"
>
          <p>
                    There is unrest in the Galactic Senate...
          </p>
          <p>
                    Attack of the Clones
          </p>
          <p>
                    2002-05-16
          </p>
</div>,
        "className": "crawl-container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <p>
              There is unrest in the Galactic Senate...
</p>,
            <p>
              Attack of the Clones
</p>,
            <p>
              2002-05-16
</p>,
          ],
          "className": "crawl-content",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "There is unrest in the Galactic Senate...",
            },
            "ref": null,
            "rendered": "There is unrest in the Galactic Senate...",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Attack of the Clones",
            },
            "ref": null,
            "rendered": "Attack of the Clones",
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "2002-05-16",
            },
            "ref": null,
            "rendered": "2002-05-16",
            "type": "p",
          },
        ],
        "type": "div",
      },
      "type": "section",
    },
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <section
          className="crawl-container"
>
          <div
                    className="crawl-content"
          >
                    <p>
                              There is unrest in the Galactic Senate...
                    </p>
                    <p>
                              Attack of the Clones
                    </p>
                    <p>
                              2002-05-16
                    </p>
          </div>
</section>,
        "className": "crawl-container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="crawl-content"
>
            <p>
                        There is unrest in the Galactic Senate...
            </p>
            <p>
                        Attack of the Clones
            </p>
            <p>
                        2002-05-16
            </p>
</div>,
          "className": "crawl-container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <p>
                There is unrest in the Galactic Senate...
</p>,
              <p>
                Attack of the Clones
</p>,
              <p>
                2002-05-16
</p>,
            ],
            "className": "crawl-content",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "There is unrest in the Galactic Senate...",
              },
              "ref": null,
              "rendered": "There is unrest in the Galactic Senate...",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Attack of the Clones",
              },
              "ref": null,
              "rendered": "Attack of the Clones",
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "2002-05-16",
              },
              "ref": null,
              "rendered": "2002-05-16",
              "type": "p",
            },
          ],
          "type": "div",
        },
        "type": "section",
      },
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
